File: 393. UTF-8 Validation.py
class Solution:
    def validUtf8(self, data: List[int]) -> bool:
        five = 0b11111000
        four = 0b11110000
        three = 0b11100000
        two = 0b11000000
        one = 0b10000000
        byte = 0b11111111
        ind = 0
        while ind < len(data):
            num = data[ind] & byte
            
            if five & num == four:
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
            elif four & num == three:
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
            elif three & num == two:
                ind += 1
                if ind >= len(data) or data[ind] & byte & two != one:
                    return False
                ind += 1
            else:
                if num & one == one:
                    return False
                ind += 1
        return True

File: 393. UTF-8 Validation.java
class Solution {
    public boolean validUtf8(int[] data) {
        for (int i=0; i<data.length; i++) {
            if ((data[i] & 0xF8) == 0xF0) {
                // 4 byte
                int count = 1;
                while (count <= 3) {
                    if (i+count >= data.length || (data[i+count] & 0xC0) != 0x80) {
                        return false;
                    }
                    count++;
                }
                i += count-1;
            } else if ((data[i] & 0xF0) == 0xE0) {
                // 3 byte
                int count = 1;
                while (count <= 2) {
                    if (i+count >= data.length || (data[i+count] & 0xC0) != 0x80) {
                        return false;
                    }
                    count++;
                }
                i += count-1;
            } else if ((data[i] & 0xE0) == 0xC0) {
                // 2 byte
                int count = 1;
                while (count <= 1) {
                    if (i+count >= data.length || (data[i+count] & 0xC0) != 0x80) {
                        return false;
                    }
                    count++;
                }
                i += count-1;
            } else if ((data[i] & 0x80) == 0) {
                // 1 byte
                continue;
            } else {
                return false;
            }
        }
        return true;
    }
}